### PR TITLE
HMR tools + miscellaneous improvements

### PR DESCRIPTION
**Pull request by maxentile**
Notable changes:
* Fixes hydrogen mass repartitioning [(HMR) logic](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/utilities/openmm_utilities.py#L36-L199), adds [tests](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/tests/test_openmm_utilities.py) of this logic, and adds [an HMR experiment](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/experiments/hmr.py)
* [Adds a barostat](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/testsystems/testsystems.py#L45) before equilibrium sampling of explicit solvent systems, and [removes it](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/testsystems/bookkeepers.py#L181) before work measurements (addressing: https://github.com/choderalab/integrator-benchmark/issues/10)
* Adds g-BAOAB experiments
  * [Delta F_neq vs. n geodesic drift steps for waterbox](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/experiments/gbaoab.py)
  * [Deviation from ideal geodesic drift vs. outer timestep, for 1-10 geodesic drift steps](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/experiments/geodesic_step_difference.py)
* Stores simulation parameters [centrally](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/__init__.py) (mostly addressing https://github.com/choderalab/integrator-benchmark/issues/11)

### CODE REVIEW

**Comment by maxentile** [comment]
If @jchodera could take a look over the [HMR implementation](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/utilities/openmm_utilities.py#L118-L199), this should be just about ready* to merge!

(*Travis tests are failing due to a conda-build error-- I'll ignore until that's fixed on their end. [These tests](https://github.com/choderalab/integrator-benchmark/blob/hmr/benchmark/tests/test_openmm_utilities.py) pass locally on Python 2.7.9 and 3.5.2.)

File: benchmark/utilities/openmm_utilities.py [hunk 1]
```diff
@@ -86,43 +114,50 @@ def get_atoms_bonded_to_hydrogen(topology):
                 atom_indices.append(b.index)
             else:
                 atom_indices.append(a.index)
-    return atom_indices
+    return list(set(atom_indices))
 
 
 def repartition_hydrogen_mass_connected(topology, system, h_mass=4.0,
                                         mode="scale"  # or "decrement"
```
Comment (jchodera) [inline benchmark/utilities/openmm_utilities.py]: Don't we want the default to be `decrement` to match what the AMBER folks do?

File: benchmark/utilities/openmm_utilities.py [hunk 2]
```diff
@@ -86,43 +114,50 @@ def get_atoms_bonded_to_hydrogen(topology):
...
-    * Should we subtract the same mass from each bonded atom?
-    * Should we proportionally reduce the mass of each bonded atom?
     """
 
-    raise (NotImplementedError())
+    others = get_atoms_bonded_to_hydrogen(topology)
+    hydrogens = get_hydrogens(topology)
+    initial_h_mass = get_sum_of_masses(system, hydrogens) / len(hydrogens)
+    initial_mass_of_others = get_sum_of_masses(system, others)
+    mass_to_remove_from_others = (h_mass - initial_h_mass) * len(hydrogens)
```
Comment (jchodera) [inline benchmark/utilities/openmm_utilities.py]: Do we really want to operate globally on the sum of all masses like this? That lumps `-CH3`, `-CH2-`, and `-CH-`, along with any `-NH-` or `-OH` cases. Is this what the algorithms really do?

Let's also cite the papers for the algorithms we implement here.

File: benchmark/utilities/openmm_utilities.py [hunk 3]
```diff
@@ -86,43 +114,50 @@ def get_atoms_bonded_to_hydrogen(topology):
...
+
+    if mode == "scale":
+        scale_factor = (initial_mass_of_others - mass_to_remove_from_others) / initial_mass_of_others
+        if scale_factor <= 0:
+            raise (RuntimeError("h_mass is too large! Can't remove this much mass from bonded atoms..."))
+
+        scale_particle_masses(system, others, scale_factor)
+
+    elif mode == "decrement":
```
Comment (jchodera) [inline benchmark/utilities/openmm_utilities.py]: I thought the AMBER approach moved a specific predetermined quantity of mass to the H from the bonded heavy atom, but this implementation seems to move a quantity of mass that depends precisely on the composition of matter in the specific system.

**Comment by maxentile** [comment]
>I've looked at the HMR algorithms you implemented, but they seem foreign to me. Can you provide the citations for the algorithms you implemented?

I've not seen the behavior of `mode="scale"` or `atoms="all"` anywhere, but those seemed natural / easy to add and compare. For `mode="decrement"`, `atoms="connected"`, you're right that I implemented this incorrectly -- I'll fix this to match the AMBER paper tomorrow morning and add references.

Also, I just found a bug with how the HMR logic interacts with the equilibrium simulator in the `hmr.py` experiment: the initial velocities are sampled assuming default masses. Will fix in the morning.

**Comment by jchodera** [comment]
Let's definitely prioritize the implementation of the most popular HMR methods that we will focus on in the paper, such as the [newer AMBER paper](http://dx.doi.org/10.1021/ct5010406) and the older [Berendsen paper](http://doi.org/10.1002/(SICI)1096-987X(199906)20:8<786::AID-JCC5>3.0.CO;2-B). 

Implementing an HMR scheme nobody has heard of that isn't cited elsewhere and using this as a model for evaluating popular HMR methods we could have easily implemented instead would greatly limit the impact of our paper.

**Comment by maxentile** [comment]
Just took a closer look at the Methods section of the AMBER HMR paper. There's not a lot of justification for the particular choice of scale factor: `"Because of technical reasons with the GPU code in Amber, we use a scaling factor of 3."`

I did a quick check of how the implemented HMR schemes affect bond vibration periods (unrealistically assuming all bonds are independent), on a couple test systems. The heuristic "objective function" I'm looking to minimize is the ratio of the largest bond vibration period to the smallest bond vibration period in the system (`max(t_i) / min(t_i)`).

Since there isn't a clear justification for `scale_factor=3`, and it doesn't appear dramatically better than `scale_factor=4` by this heuristic, I would favor looking adding the AMBER scheme with `scale_factor=2,3,4` to the benchmark.

```
AlanineDipeptideVacuum
	Default max(t_i) / min(t_i)       : 2.990
	AMBER (2x)                        : 2.122	(Improvement over default: 1.409x)
	AMBER (3x)                        : 1.758	(Improvement over default: 1.700x)
	AMBER (4x)                        : 1.801	(Improvement over default: 1.660x)
	decrement (all) H mass=2.0        : 2.092	(Improvement over default: 1.429x)
	decrement (all) H mass=3.0        : 1.686	(Improvement over default: 1.773x)
	decrement (all) H mass=4.0        : 1.440	(Improvement over default: 2.076x)
	decrement (connected) H mass=2.0  : 2.089	(Improvement over default: 1.432x)
	decrement (connected) H mass=3.0  : 1.689	(Improvement over default: 1.770x)
	decrement (connected) H mass=4.0  : 1.435	(Improvement over default: 2.084x)
	scale (all) H mass=2.0            : 2.104	(Improvement over default: 1.421x)
	scale (all) H mass=3.0            : 1.702	(Improvement over default: 1.757x)
	scale (all) H mass=4.0            : 1.460	(Improvement over default: 2.049x)
	scale (connected) H mass=2.0      : 2.097	(Improvement over default: 1.426x)
	scale (connected) H mass=3.0      : 1.710	(Improvement over default: 1.748x)
	scale (connected) H mass=4.0      : 1.479	(Improvement over default: 2.021x)
AlanineDipeptideExplicit
	Default max(t_i) / min(t_i)       : 4.610
	AMBER (2x)                        : 3.152	(Improvement over default: 1.462x)
	AMBER (3x)                        : 2.509	(Improvement over default: 1.838x)
	AMBER (4x)                        : 2.109	(Improvement over default: 2.186x)
	decrement (all) H mass=2.0        : 2.997	(Improvement over default: 1.538x)
	decrement (all) H mass=3.0        : 2.211	(Improvement over default: 2.085x)
	decrement (all) H mass=4.0        : 1.684	(Improvement over default: 2.737x)
	decrement (connected) H mass=2.0  : 3.087	(Improvement over default: 1.493x)
	decrement (connected) H mass=3.0  : 2.365	(Improvement over default: 1.949x)
	decrement (connected) H mass=4.0  : 1.871	(Improvement over default: 2.464x)
	scale (all) H mass=2.0            : 3.064	(Improvement over default: 1.505x)
	scale (all) H mass=3.0            : 2.317	(Improvement over default: 1.990x)
	scale (all) H mass=4.0            : 1.833	(Improvement over default: 2.516x)
	scale (connected) H mass=2.0      : 3.128	(Improvement over default: 1.474x)
	scale (connected) H mass=3.0      : 2.447	(Improvement over default: 1.884x)
	scale (connected) H mass=4.0      : 2.008	(Improvement over default: 2.296x)
DHFRExplicit
	Default max(t_i) / min(t_i)       : 6.499
	AMBER (2x)                        : 4.291	(Improvement over default: 1.514x)
	AMBER (3x)                        : 3.208	(Improvement over default: 2.026x)
	AMBER (4x)                        : 2.465	(Improvement over default: 2.637x)
	decrement (all) H mass=2.0        : 4.308	(Improvement over default: 1.509x)
	decrement (all) H mass=3.0        : 3.236	(Improvement over default: 2.008x)
	decrement (all) H mass=4.0        : 2.522	(Improvement over default: 2.577x)
	decrement (connected) H mass=2.0  : 4.321	(Improvement over default: 1.504x)
	decrement (connected) H mass=3.0  : 3.243	(Improvement over default: 2.004x)
	decrement (connected) H mass=4.0  : 2.509	(Improvement over default: 2.590x)
	scale (all) H mass=2.0            : 4.335	(Improvement over default: 1.499x)
	scale (all) H mass=3.0            : 3.294	(Improvement over default: 1.973x)
	scale (all) H mass=4.0            : 2.623	(Improvement over default: 2.478x)
	scale (connected) H mass=2.0      : 4.393	(Improvement over default: 1.479x)
	scale (connected) H mass=3.0      : 3.382	(Improvement over default: 1.922x)
	scale (connected) H mass=4.0      : 2.725	(Improvement over default: 2.385x)```

**Comment by maxentile** [comment]
Tests pass, merging

**Pull request merged**